### PR TITLE
refactor(dirname): implement pure string manipulation per POSIX

### LINKED ISSUE #8924: dirname fails with some class of paths
> There are still incompatibilities not addressed by #8911.
> 
> | command | GNU | uutils |
> | --- | --- | --- |
> | `dirname foo//.` | `foo` | `foo/` |
> | `dirname foo///.` | `foo` | `foo//` |
> | `dirname foo/./` | `foo` | `.` |
> | `dirname foo/bar/./` | `foo/bar` | `foo` |
> | `dirname foo/./bar` | `foo/.` | `foo` |
>  

 _Originally posted by @andersk in [#8910](https://github.com/uutils/coreutils/issues/8910#issuecomment-3408735720)_

### PR DESCRIPTION
**Pull request by naoNao89**
Fixes #8924 by rewriting dirname as pure byte-level string operations, avoiding Path::parent() normalization.

Handles all edge cases: foo//., foo///., foo/./, etc. Implementation uses explicit step-by-step logic without unsafe code, making it maintainable and platform-safe.

Added comprehensive tests for slash variations and dot-slash patterns.

Alternative to #8927 - this approach prioritizes code clarity and safety over conciseness. Both fix the same issues, but this avoids unsafe assumptions about OsStr encoding.

### CODE REVIEW

**Comment by sylvestre** [comment]
@naoNao89 please don't start new pr when others started. Contribute to the other instead :)

File: tests/by-util/test_dirname.rs [hunk 1]
```diff
@@ -216,3 +223,68 @@ fn test_existing_behavior_preserved() {
...
+        .succeeds()
+        .stdout_is("/home/dos\n/var\n.\n/tmp\n.\n/\nrelative\n");
+}
+
+#[test]
+fn test_all_dot_slash_variations() {
+    // Tests for all the cases mentioned in issue #8910 comment
+    // https://github.com/uutils/coreutils/issues/8910#issuecomment-3408735720
+
+    // foo//. -> foo
```
Comment (sylvestre) [inline tests/by-util/test_dirname.rs]: please remove the comments, they don't provide much here. reading the test is as easy

File: tests/by-util/test_dirname.rs [hunk 2]
```diff
@@ -216,3 +223,68 @@ fn test_existing_behavior_preserved() {
...
         .stdout_is("/home/dos\n");
 }
+
+#[test]
+fn test_multiple_paths_comprehensive() {
+    // Comprehensive test for multiple paths in single invocation
+    // Tests the loop at line 84 with various edge cases mixed
+    new_ucmd!()
+        .args(&[
+            "/home/dos/.",   // trailing dot case
```
Comment (sylvestre) [inline tests/by-util/test_dirname.rs]: same
, please remove comments

File: tests/by-util/test_dirname.rs [hunk 3]
```diff
@@ -216,3 +223,68 @@ fn test_existing_behavior_preserved() {
         .succeeds()
         .stdout_is("/home/dos\n");
 }
+
+#[test]
+fn test_multiple_paths_comprehensive() {
+    // Comprehensive test for multiple paths in single invocation
+    // Tests the loop at line 84 with various edge cases mixed
```
Comment (sylvestre) [inline tests/by-util/test_dirname.rs]: line 84 ? please remove this

**Comment by sylvestre** [comment]
why draft ? :)

**Comment by RenjiSann** [comment]
@naoNao89 what's the status on this ? It has conflicts

**Comment by naoNao89** [comment]
resolved

**Comment by naoNao89** [comment]
Caught an `.unwrap()` crash on an unrelated test where a process was killed by signal instead of exiting normally 💀 #9430

File: tests/by-util/test_dirname.rs [hunk 4]
```diff
@@ -9,6 +9,13 @@ fn test_invalid_arg() {
     new_ucmd!().arg("--definitely-invalid").fails_with_code(1);
 }
 
+#[test]
+fn test_missing_operand() {
+    // Test calling dirname with no arguments - should fail
+    // This covers the error path at line 80-82 in dirname.rs
```
Comment (sylvestre) [inline tests/by-util/test_dirname.rs]:
```suggestion
```

lines will change

File: src/uu/dirname/src/dirname.rs [hunk 5]
```diff
@@ -18,51 +18,84 @@ mod options {
...
 ///
-/// dirname should do simple string manipulation without path normalization.
 /// See issue #8910 and similar fix in basename (#8373, commit c5268a897).
-///
-/// Returns `Some(())` if the special case was handled (output already printed),
-/// or `None` if normal `Path::parent()` logic should be used.
-fn handle_trailing_dot(path_bytes: &[u8]) -> Option<()> {
-    if !path_bytes.ends_with(b"/.") {
-        return None;
+fn dirname_string_manipulation(path_bytes: &[u8]) -> Vec<u8> {
```
Comment (sylvestre) [inline src/uu/dirname/src/dirname.rs]: should return a Cow<[u8]>  ?

**Comment by sylvestre** [comment]
follow up https://github.com/uutils/coreutils/pull/10294
